test/vfd_swmr_group_writer.c

```diff
@@ -112,216 +157,2736 @@ state_init(state_t *s, int argc, char **argv)
+ *
+ *              unsigned int which
+ *              The number of iterations for group creation, use to generate 
+ *              newly created group and attribute names. 
+ *
+ * Return:      Success:    true
+ *              Failure:    false
+ *
+ * Note:        This is an internal function used by the 
+ *              "dense-del-to-compact" test. 
+ *              For attribute compact/dense storage, check the reference
+ *              manual of H5Pget_attr_phase_change.
+ *-------------------------------------------------------------------------
+*/ 
+
+static bool
+del_attrs_compact_dense_compact(state_t *s, 
+                                hid_t obj_id,
+                                hid_t gcpl,
+                                unsigned int which) {
+    
+    unsigned max_compact = 0;
+    unsigned min_dense = 0;
+    unsigned u = 0;
+
+    char attrname[VS_ATTR_NAME_LEN];
+    const char* aname_format="attr-%u-%u";
+    const char* adname_format="attr-d-%u-%u";
+
+    /* Obtain the maximal number of attributes to be stored in compact
+     * storage and the minimal number of attributes to be stored in
+     * dense storage. */
+    if (H5Pget_attr_phase_change(gcpl, &max_compact, &min_dense) < 0) {
+        H5_FAILED(); AT();
+        printf("H5Pget_attr_phase_change failed\n");
+        goto error;
+    }
+    u= max_compact +1;
+
+
+    /* delete a number of attributes so that the attribute storage just becomes dense.*/
+    for(u--;u>=(min_dense-1);u--) {
+        HDsprintf(attrname, aname_format, which,max_compact-u);
+        if (H5Adelete(obj_id,attrname) < 0) {
+            H5_FAILED(); AT();
+            printf("H5Adelete failed\n");
+            goto error;
+        }
+
+        /* For each attribute deletion, we want to ensure the verification
+         * from the reader. 
+         * So writer sends a message to reader: an attribute is successfully deleted. 
+           then wait for reader to verify and send an acknowledgement message back. */
+        if(s->use_named_pipes && s->attr_test == true) {
+            dbgf(2, "writer: delete attr - ready to send the message: %d\n", s->np_notify+1);
+            if(np_wr_send_receive(s) == false) {
+                H5_FAILED(); AT();
+                dbgf(2, "writer: delete attr - verification failed.\n");
+                goto error2;
+            }
+        }
+    }
+
+    /* The writer deletes another attribute, the storage is
+     * still dense. However, the attribute to be deleted
+     * doesn't follow the previous for loop. It may be 
+     * in different location in the object header. Just add
+     * a litter variation to check if this operation is successful. 
+     * The attribute name to be deleted is attr-max_compact+which-0
+     */
+     
+    HDsprintf(attrname,adname_format,max_compact+which,0);
+    if (H5Adelete(obj_id,attrname) < 0) {
+            H5_FAILED(); AT();
+            printf("H5Adelete failed\n");
```

Comment:
We have mixed ways of printing out error messages.  Sometimes `printf`, sometimes `dbgf`.  When we have time, we should clean them up.

Reply:
dbgf is not for printing out error messages. It is just for debugging. Agree that we may make the dbgf consistent.